provers/groth16/src/qap.rs

```diff
@@ -125,3 +65,100 @@ impl QuadraticArithmeticProgram {
         })
     }
 }
+
+impl QuadraticArithmeticProgram {
+    pub fn from_r1cs(r1cs: R1CS) -> QuadraticArithmeticProgram {
+        let num_gates = r1cs.number_of_constraints();
+        let next_power_of_two = num_gates.next_power_of_two();
+        let pad_zeroes = next_power_of_two - num_gates;
+
+        let mut l: Vec<Polynomial<FrElement>> = vec![];
+        let mut r: Vec<Polynomial<FrElement>> = vec![];
+        let mut o: Vec<Polynomial<FrElement>> = vec![];
+        for i in 0..r1cs.witness_size() {
+            let [l_poly, r_poly, o_poly] =
+                get_variable_lro_polynomials_from_r1cs(&r1cs, i, pad_zeroes);
+            l.push(l_poly);
+            r.push(r_poly);
+            o.push(o_poly);
+        }
+
+        QuadraticArithmeticProgram {
+            l,
+            r,
+            o,
+            num_of_gates: next_power_of_two,
+            num_of_public_inputs: r1cs.number_of_inputs + 1, // Compansate for "1"
+        }
+    }
+
+    pub fn from_variable_matrices(
+        num_of_public_inputs: usize,
+        l: &[Vec<FrElement>],
+        r: &[Vec<FrElement>],
+        o: &[Vec<FrElement>],
+    ) -> Self {
+        let num_of_total_inputs = l.len();
+        assert_eq!(num_of_total_inputs, r.len());
+        assert_eq!(num_of_total_inputs, o.len());
+        assert!(num_of_total_inputs > 0);
+        assert!(num_of_public_inputs <= num_of_total_inputs);
```

Comment:
This can panic, it should return an Error

Reply:
This method is only for testing purposes. I moved the method to an utils.rs file within the tests.

Reply:
If that's the case you can add the conditional compile flag, the cfg tests. Or just move it to the test module